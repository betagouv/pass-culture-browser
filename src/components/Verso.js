import classnames from 'classnames'
import React, { Component } from 'react'
import { connect } from 'react-redux'

import OfferInfo from './OfferInfo'
import VersoWrapper from './VersoWrapper'
import MenuButton from './layout/MenuButton'
import selectCurrentMediation from '../selectors/currentMediation'
import selectIsCurrentTuto from '../selectors/isCurrentTuto'
import { THUMBS_URL } from '../utils/config'

class Verso extends Component {
  render() {
    const {
      mediation,
      isFlipped,
      isCurrentTuto
    } = this.props
    return (
      <div
        className={classnames('verso', {
          flipped: isFlipped,
        })}
      >
        <VersoWrapper
          hasControlBar={!isCurrentTuto}
          className="with-padding-top"
        >
          {isCurrentTuto ? (
            mediation && (
              <img
                alt="verso"
                className="verso-tuto-mediation"
                src={`${THUMBS_URL}/mediations/${mediation.id}_1`}
              />
            )
          ) : (
            <OfferInfo />
          )}
        </VersoWrapper>
        {isCurrentTuto ? (
          <MenuButton borderTop />
        ) : (
          <MenuButton borderTop colored />
        )}
      </div>
    )
  }
}

export default connect(state => ({
  mediation: selectCurrentMediation(state),
  isFlipped: state.verso.isFlipped,
  isCurrentTuto: selectIsCurrentTuto(state),
}))(Verso)
