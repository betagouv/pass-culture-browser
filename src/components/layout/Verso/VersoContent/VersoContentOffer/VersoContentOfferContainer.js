import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'
import { requestData } from 'redux-thunk-data'

import {
  selectBookables,
  selectBookingByRouterMatch
} from '../../../../../selectors/data/bookingsSelectors'
import { selectOfferByRouterMatch } from '../../../../../selectors/data/offersSelectors'
import { selectDistanceByRouterMatch } from '../../../../../selectors/geolocationSelectors'
import { selectIsNotBookableByRouterMatch } from '../../../../../selectors/isNotBookableSelector'
import { selectUserGeolocation } from '../../../../../selectors/geolocationSelectors'
import getStyle from './utils/getStyle'
import VersoContentOffer from './VersoContentOffer'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const offer = selectOfferByRouterMatch(state, match) || {}
  const bookables = selectBookables(state, offer)
  const isNotBookable = selectIsNotBookableByRouterMatch(state, match)
  const { extraData } = offer || {}
  const style = getStyle(state, extraData)
  const booking = selectBookingByRouterMatch(state, match)
  const distance = selectDistanceByRouterMatch(state, match)
  const userGeolocation = selectUserGeolocation(state)
  const isCancelled = booking && booking.isCancelled

  return {
    bookables,
    booking,
    distance,
    isCancelled,
    isNotBookable,
    offer,
    style,
    userGeolocation
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  handleRequestMusicAndShowTypes: () => {
    const { musicTypes, showTypes } = ownProps

    if (!musicTypes) {
      dispatch(
        requestData({
          apiPath: '/musicTypes'
        })
      )
    }

    if (!showTypes) {
      dispatch(
        requestData({
          apiPath: '/showTypes'
        })
      )
    }
  }
})

export default compose(
  withRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(VersoContentOffer)
