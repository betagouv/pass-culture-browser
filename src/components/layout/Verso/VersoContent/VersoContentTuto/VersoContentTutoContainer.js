import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'

import VersoContentTuto from './VersoContentTuto'
import { selectThumbUrlByRouterMatch } from '../../../../../selectors/data/thumbUrlSelector'

const VERSO_OVH_THUMB_SUFFIX = '_1'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const thumbUrl = selectThumbUrlByRouterMatch(state, match)
  return {
    imageURL: `${thumbUrl}${VERSO_OVH_THUMB_SUFFIX}`,
  }
}

export default compose(
  withRouter,
  connect(mapStateToProps)
)(VersoContentTuto)
