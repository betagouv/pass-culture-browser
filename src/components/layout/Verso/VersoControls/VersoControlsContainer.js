import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'

import VersoControls from './VersoControls'
import selectBookingByRouterMatch from '../../../../selectors/selectBookingByRouterMatch'

const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const booking = selectBookingByRouterMatch(state, match)
  const showCancelView = booking && !booking.isCancelled
  return {
    showCancelView,
  }
}

export default compose(
  withRouter,
  connect(mapStateToProps)
)(VersoControls)
