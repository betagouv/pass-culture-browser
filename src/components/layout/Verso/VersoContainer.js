import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'

import Verso from './Verso'
import getHeaderColor from '../../../utils/colors'
import selectMediationByRouterMatch from '../../../selectors/selectMediationByRouterMatch'
import selectOfferByRouterMatch from '../../../selectors/selectOfferByRouterMatch'
import { ROOT_PATH } from '../../../utils/config'

const backgroundImage = `url('${ROOT_PATH}/mosaic-k.png')`

export const checkIsTuto = mediation => {
  const { tutoIndex } = mediation || {}
  const result = Boolean(typeof tutoIndex === 'number')
  return result
}

export const getContentInlineStyle = (isTuto, backgroundColor) => {
  let result = { backgroundImage }
  if (isTuto && backgroundColor) result = { ...result, backgroundColor }
  return result
}

const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const offer = selectOfferByRouterMatch(state, match) || {}
  const mediation = selectMediationByRouterMatch(state, match) || {}
  const firstThumbDominantColor = offer.firstThumbDominantColor || mediation.firstThumbDominantColor
  const backgroundColor = getHeaderColor(firstThumbDominantColor)
  const isTuto = checkIsTuto(mediation)
  const contentInlineStyle = getContentInlineStyle(isTuto, backgroundColor)

  const { name: offerName, venue } = offer
  const { name: venueName, publicName: venuePublicName } = venue || {}
  const offerVenueNameOrPublicName = venuePublicName || venueName

  return {
    backgroundColor,
    contentInlineStyle,
    isTuto,
    offerName,
    offerVenueNameOrPublicName,
  }
}

export default compose(
  withRouter,
  connect(mapStateToProps)
)(Verso)
