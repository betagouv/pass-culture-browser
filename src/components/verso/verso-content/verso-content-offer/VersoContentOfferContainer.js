import get from 'lodash.get'
import { connect } from 'react-redux'
import { requestData } from 'redux-saga-data'

import VersoContentOffer from './VersoContentOffer'
import selectMusicTypeByCode from './selectors/selectMusicTypeByCode'
import selectMusicSubTypeByCodeAndSubCode from './selectors/selectMusicSubTypeByCodeAndSubCode'
import selectShowTypeByCode from './selectors/selectShowTypeByCode'
import selectShowSubTypeByCodeAndSubCode from './selectors/selectShowSubTypeByCodeAndSubCode'
import { isRecommendationOfferFinished } from '../../../../helpers'
import selectBookables from '../../../../selectors/selectBookables'

export const mapStateToProps = (state, ownProps) => {
  const { recommendation } = ownProps
  const { offerId } = recommendation

  const bookables = selectBookables(state, recommendation)
  const isFinished = isRecommendationOfferFinished(recommendation, offerId)

  const extraData = get(recommendation, 'offer.product.extraData')
  const musicType = selectMusicTypeByCode(state, get(extraData, 'musicType'))
  const showType = selectShowTypeByCode(state, get(extraData, 'showType'))
  const musicSubType = selectMusicSubTypeByCodeAndSubCode(
    state,
    get(extraData, 'musicType'),
    get(extraData, 'musicSubType')
  )
  const showSubType = selectShowSubTypeByCodeAndSubCode(
    state,
    get(extraData, 'showType'),
    get(extraData, 'showSubType')
  )

  return {
    bookables,
    isFinished,
    musicSubType,
    musicType,
    musicTypes: get(state, 'data.musicTypes'),
    showSubType,
    showType,
    showTypes: get(state, 'data.showTypes'),
  }
}

const mapDispatchToProps = (dispatch, ownProps) => {
  const { musicTypes, showTypes } = ownProps
  return {
    handleRequestMusicAndShowTypes: () => {
      if (!musicTypes) {
        dispatch(
          requestData({
            apiPath: '/musicTypes',
          })
        )
      }
      if (!showTypes) {
        dispatch(
          requestData({
            apiPath: '/showTypes',
          })
        )
      }
    },
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(VersoContentOffer)
