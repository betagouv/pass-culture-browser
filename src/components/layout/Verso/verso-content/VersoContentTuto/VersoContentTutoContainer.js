import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { compose } from 'redux'

import VersoContentTuto from './VersoContentTuto'
import selectThumbUrlByRouterMatch from '../../../../../selectors/selectThumbUrlByRouterMatch'

const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const thumbUrl = selectThumbUrlByRouterMatch(state, match)
  return {
    imageURL: thumbUrl,
  }
}

export default compose(
  withRouter,
  connect(mapStateToProps)
)(VersoContentTuto)
