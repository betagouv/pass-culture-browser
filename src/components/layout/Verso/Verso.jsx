import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import VersoContentOfferContainer from './verso-content/VersoContentOffer/VersoContentOfferContainer'
import VersoContentTutoContainer from './verso-content/VersoContentTuto/VersoContentTutoContainer'
import VersoControlsContainer from './VersoControls/VersoControlsContainer'
import VersoHeaderContainer from './verso-content/VersoHeaderContainer'
import AbsoluteFooterContainer from '../AbsoluteFooter/AbsoluteFooterContainer'

class Verso extends React.PureComponent {
  constructor(props) {
    super(props)
    this.versoWrapper = React.createRef()
  }

  componentDidUpdate() {
    this.versoWrapper.current.scrollTo(0, 0)
  }

  render() {
    const {
      areDetailsVisible,
      backgroundColor,
      contentInlineStyle,
      extraClassName,
      isTuto,
      offerName,
      offerVenueNameOrPublicName,
    } = this.props

    return (
      <div
        className={classnames('verso', extraClassName, {
          flipped: areDetailsVisible,
        })}
      >
        <div
          className="verso-wrapper"
          ref={this.versoWrapper}
        >
          <VersoHeaderContainer
            backgroundColor={backgroundColor}
            subtitle={offerVenueNameOrPublicName}
            title={offerName}
          />
          {!isTuto && <VersoControlsContainer />}
          <div
            className="verso-content"
            style={contentInlineStyle}
          >
            {isTuto ? <VersoContentTutoContainer /> : <VersoContentOfferContainer />}
          </div>
        </div>
        <AbsoluteFooterContainer
          areDetailsVisible={areDetailsVisible}
          borderTop
          colored={!isTuto}
          id="verso-footer"
        />
      </div>
    )
  }
}

Verso.defaultProps = {
  backgroundColor: null,
  contentInlineStyle: null,
  extraClassName: null,
  isTuto: null,
  offerName: null,
  offerVenueNameOrPublicName: null,
}

Verso.propTypes = {
  areDetailsVisible: PropTypes.bool.isRequired,
  backgroundColor: PropTypes.string,
  contentInlineStyle: PropTypes.shape(),
  extraClassName: PropTypes.string,
  isTuto: PropTypes.bool,
  offerName: PropTypes.string,
  offerVenueNameOrPublicName: PropTypes.string,
}

export default Verso
