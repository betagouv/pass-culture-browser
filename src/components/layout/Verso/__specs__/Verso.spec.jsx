import React from 'react'
import { shallow } from 'enzyme'

import Verso from '../Verso'
import VersoContentOfferContainer from '../verso-content/VersoContentOffer/VersoContentOfferContainer'
import VersoContentTutoContainer from '../verso-content/VersoContentTuto/VersoContentTutoContainer'
import VersoControlsContainer from '../VersoControls/VersoControlsContainer'

const backgroundColor = '#ACE539'
const props = {
  areDetailsVisible: true,
  backgroundColor,
  contentInlineStyle: { backgroundColor, backgroundImage: 'any/image' },
  extraClassName: 'extra-classname',
  isTuto: false,
  mediationId: 'AAA',
  offerName: 'Offer title',
  offerVenueNameOrPublicName: 'Offer subtitle',
}

describe('src | components | verso | Verso', () => {
  it('should match snapshot', () => {
    // when
    const wrapper = shallow(<Verso {...props} />)

    // then
    expect(wrapper).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('should show offer view when is not tuto', () => {
    // when
    const wrapper = shallow(<Verso {...props} />)
    const infos = wrapper.find(VersoContentOfferContainer)
    const tuto = wrapper.find(VersoContentTutoContainer)
    const controls = wrapper.find(VersoControlsContainer)

    // then
    expect(tuto).toHaveLength(0)
    expect(infos).toHaveLength(1)
    expect(controls).toHaveLength(1)
  })

  it('should show tuto view when is tuto', () => {
    // given
    props.isTuto = true

    // when
    const wrapper = shallow(<Verso {...props} />)
    const infos = wrapper.find(VersoContentOfferContainer)
    const tuto = wrapper.find(VersoContentTutoContainer)
    const controls = wrapper.find(VersoControlsContainer)

    // then
    expect(tuto).toHaveLength(1)
    expect(infos).toHaveLength(0)
    expect(controls).toHaveLength(0)
  })
})
