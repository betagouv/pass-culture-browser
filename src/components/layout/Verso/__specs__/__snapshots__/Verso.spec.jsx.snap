// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | Verso should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Verso
    areDetailsVisible={true}
    backgroundColor="#ACE539"
    contentInlineStyle={
      Object {
        "backgroundColor": "#ACE539",
        "backgroundImage": "any/image",
      }
    }
    extraClassName="extra-classname"
    isTuto={false}
    offerName="Offer title"
    offerVenueNameOrPublicName="Offer subtitle"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="verso-wrapper"
        >
          <withRouter(withSizes(VersoHeader))
            backgroundColor="#ACE539"
            subtitle="Offer subtitle"
            title="Offer title"
          />
          <withRouter(Connect(VersoControls)) />
          <div
            className="verso-content"
            style={
              Object {
                "backgroundColor": "#ACE539",
                "backgroundImage": "any/image",
              }
            }
          >
            <withRouter(Connect(VersoContentOffer)) />
          </div>
        </div>,
        <withRouter(AbsoluteFooter)
          areDetailsVisible={true}
          borderTop={true}
          colored={true}
          id="verso-footer"
        />,
      ],
      "className": "verso extra-classname flipped",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <withRouter(withSizes(VersoHeader))
              backgroundColor="#ACE539"
              subtitle="Offer subtitle"
              title="Offer title"
            />,
            <withRouter(Connect(VersoControls)) />,
            <div
              className="verso-content"
              style={
                Object {
                  "backgroundColor": "#ACE539",
                  "backgroundImage": "any/image",
                }
              }
            >
              <withRouter(Connect(VersoContentOffer)) />
            </div>,
          ],
          "className": "verso-wrapper",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "backgroundColor": "#ACE539",
              "subtitle": "Offer subtitle",
              "title": "Offer title",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <withRouter(Connect(VersoContentOffer)) />,
              "className": "verso-content",
              "style": Object {
                "backgroundColor": "#ACE539",
                "backgroundImage": "any/image",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "areDetailsVisible": true,
          "borderTop": true,
          "colored": true,
          "id": "verso-footer",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="verso-wrapper"
          >
            <withRouter(withSizes(VersoHeader))
              backgroundColor="#ACE539"
              subtitle="Offer subtitle"
              title="Offer title"
            />
            <withRouter(Connect(VersoControls)) />
            <div
              className="verso-content"
              style={
                Object {
                  "backgroundColor": "#ACE539",
                  "backgroundImage": "any/image",
                }
              }
            >
              <withRouter(Connect(VersoContentOffer)) />
            </div>
          </div>,
          <withRouter(AbsoluteFooter)
            areDetailsVisible={true}
            borderTop={true}
            colored={true}
            id="verso-footer"
          />,
        ],
        "className": "verso extra-classname flipped",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <withRouter(withSizes(VersoHeader))
                backgroundColor="#ACE539"
                subtitle="Offer subtitle"
                title="Offer title"
              />,
              <withRouter(Connect(VersoControls)) />,
              <div
                className="verso-content"
                style={
                  Object {
                    "backgroundColor": "#ACE539",
                    "backgroundImage": "any/image",
                  }
                }
              >
                <withRouter(Connect(VersoContentOffer)) />
              </div>,
            ],
            "className": "verso-wrapper",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "backgroundColor": "#ACE539",
                "subtitle": "Offer subtitle",
                "title": "Offer title",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <withRouter(Connect(VersoContentOffer)) />,
                "className": "verso-content",
                "style": Object {
                  "backgroundColor": "#ACE539",
                  "backgroundImage": "any/image",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "areDetailsVisible": true,
            "borderTop": true,
            "colored": true,
            "id": "verso-footer",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
